src/connection/activate.ts

```diff
@@ -0,0 +1,94 @@
+import { sendAnalyticsEvent } from '@uniswap/analytics'
+import { InterfaceEventName, WalletConnectionResult } from '@uniswap/analytics-events'
+import { sendEvent } from 'components/analytics'
+import { Connection } from 'connection'
+import { atom, useAtom } from 'jotai'
+import { useCallback } from 'react'
+import { useAppDispatch } from 'state/hooks'
+import { updateSelectedWallet } from 'state/user/reducer'
+
+import { didUserReject } from './utils'
+
+type PendingConnectionState = {
+  connection: Connection | undefined
+  error: any
+}
+
+const pendingConnectionAtom = atom<PendingConnectionState>({ connection: undefined, error: undefined })
+
+export function useActivateConnection() {
+  const dispatch = useAppDispatch()
+
+  const [pending, setPending] = useAtom(pendingConnectionAtom)
+
+  const updatePendingConnection = useCallback(
+    (connection: Connection | undefined) => {
+      // log selected wallet
+      if (connection !== undefined) {
+        sendEvent({
+          category: 'Wallet',
+          action: 'Change Wallet',
+          label: connection.type,
+        })
+      }
+      // Setting a new connection clears any pending errors
+      setPending({ connection, error: undefined })
+    },
+    [setPending]
+  )
+
+  const updatePendingError = useCallback(
+    (error: any) => {
+      setPending(({ connection }) => {
+        if (error !== undefined && connection !== undefined) {
```

Comment:
Why `if (error !== undefined && connection !== undefined) {` here? Shouldn't the caller make sure to call `updatePendingError` when there's an error, so we don't need to check here for its presence and noop otherwise?

Reply:
Good catch, I don't think I actually needed this check for error at all, removed it